packages/venia-concept/src/components/SwatchTooltip/swatchTooltip.js

```diff
@@ -0,0 +1,110 @@
+import React, { Component } from 'react';
+import { shape, string } from 'prop-types';
+import over from 'lodash.over';
+import uuid from 'uuid/v4';
+
+import classify from 'src/classify';
+import defaultClasses from './swatchTooltip.css';
+
+class SwatchTooltip extends Component {
+    static propTypes = {
+        classes: shape({
+            root: string,
+            tooltip: string
+        }),
+        text: string
+    };
+
+    constructor(...args) {
+        super(...args);
+
+        this.timeoutId = null;
+        this.uniqueId = uuid();
+    }
+
+    state = {
+        isShowing: false
+    };
+
+    /*
+     *  Close the tooltip on the next tick.
+     *  Unfortunately this is necessary because another child of ours may have received focus
+     *  and this blur event is fired before that new focus event.
+     *  @see https://reactjs.org/docs/accessibility.html#mouse-and-pointer-events.
+     */
+    onBlur = () => {
+        this.timeoutId = setTimeout(() => {
+            this.setState({ isShowing: false });
+        }, 0);
+    };
+
+    /*
+     *  The tooltip target has received focus, show the tooltip.
+     */
+    onFocus = () => {
+        // If a child receives focus, do not close the tooltip yet.
+        clearTimeout(this.timeoutId);
+
+        this.setState({ isShowing: true });
+    };
+
+    onKeyDown = event => {
+        if (event.key === 'Escape') {
+            this.setState({ isShowing: false });
+        }
+    };
+    onMouseOver = () => {
+        this.setState({ isShowing: true });
+    };
+    onMouseLeave = () => {
+        this.setState({ isShowing: false });
+    };
+
+    render() {
+        const { text, children, classes } = this.props;
+        const { isShowing } = this.state;
+
+        const ariaEnhancedChildren = React.Children.map(children, child => {
+            if (child.type !== 'button') {
+                return child;
+            }
+
+            // The button triggers the tooltip.
+            // Update its accessibility accordingly.
+            const ariaEnhancedButton = React.cloneElement(child, {
+                'aria-describedby': this.uniqueId,
+                // We want to tack our event functionality onto whatever already exists.
+                // lodash.over returns a function that calls each function
+                // provided to it with the args it receives -
+                // in this case, a React SyntheticEvent.
+                onBlur: over([child.props.onBlur, this.onBlur]),
+                onFocus: over([child.props.onFocus, this.onFocus]),
+                onKeyDown: over([child.props.onKeyDown, this.onKeyDown]),
+                onMouseOver: over([child.props.onMouseOver, this.onMouseOver]),
+                onMouseLeave: over([
+                    child.props.onMouseLeave,
+                    this.onMouseLeave
+                ])
```

Comment:
I've never seen this pattern before. Ordinarily, you would call the props callback from the class handler. Any reason we shouldn't do that here?

Also, by convention, class handlers are named `handleEvent` rather than `onEvent`.

```jsx
class Foo extends Component {
  handleClick = event => {
    const { id, onClick } = this.props

    // this component's business
    event.preventDefault()

    // its parent's business
    // note: event object is treated as private
    // callback gets only what it needs
    if (typeof onClick === "function") {
      onClick(id)
    }
  }

  render() {
    const { handleClick } = this

    return <button onClick={handleClick} />
  }
}
```

Reply:
Yeah that's a good call - the reason this particular case is different is because this `SwatchTooltip` component is the parent, not the child.

In other words:

```jsx
<SwatchTooltip>
  <Swatch />
</SwatchTooltip>
```

So for a given event handler - let's say `onClick` - `SwatchTooltip` (the parent) wants its handler executed while also not trampling on `Swatch` (the child)'s handler.

I suppose I could do something like

```js
    onMouseOver = () => {
        // parent business.
        this.setState({ isShowing: true });

        // child business.
        const child = this.getChild(); // some logic that uses React.Children to get the right one
        if (child.props.onMouseOver) {
          child.props.onMouseOver();
        }
    };
```

and repeat that pattern for each event handler.

I'm mostly just trying to avoid tightly coupling these components.

Reply:
> I'm mostly just trying to avoid tightly coupling these components.

Why's that? It seems to me that the tooltip should be a child of the `Swatch`, rendered by `Swatch`, and pretty coupled to it.

Reply:
> Why's that? It seems to me that the tooltip should be a child of the Swatch, rendered by Swatch, and pretty coupled to it.

Good point, that would definitely make things much simpler here. In the future we may want a generic `Tooltip` component like this, but that's out of scope here.